src/stream.rs

```diff
@@ -35,6 +35,59 @@ pub enum Stream {
     Test(Box<dyn Read + Send>, Vec<u8>),
 }
 
+// DeadlineStream wraps a stream such that read() will return an error
+// after the provided deadline, and sets timeouts on the underlying
+// TcpStream to ensure read() doesn't block beyond the deadline.
+// When the From trait is used to turn a DeadlineStream back into a
+// Stream (by PoolReturningRead), the timeouts are removed.
+pub struct DeadlineStream {
+    stream: Stream,
+    deadline: Option<Instant>,
+}
+
+impl DeadlineStream {
+    pub(crate) fn new(stream: Stream, deadline: Option<Instant>) -> Self {
+        DeadlineStream { stream, deadline }
+    }
+}
+
+impl From<DeadlineStream> for Stream {
+    fn from(deadline_stream: DeadlineStream) -> Stream {
+        // Since we are turning this back into a regular, non-deadline Stream,
+        // remove any timeouts we set.
+        let stream = deadline_stream.stream;
+        if let Some(socket) = stream.socket() {
+            socket.set_read_timeout(None).unwrap();
+            socket.set_write_timeout(None).unwrap();
+        }
+        stream
+    }
+}
```

Comment:
The rust idiomatic way for returning a wrapped value is often called `into_inner`. It's not based on traits or anything, so it isn't enforced, it's just very common to see both in the std lib and in other crates.

[`BufReader`](https://doc.rust-lang.org/std/io/struct.BufReader.html#method.into_inner) is a relevant example.

Since `DeadlineStream` is internal to ureq, I don't really care if you change or keep it like this. Just worth noting.

Reply:
Good point. I like that idiom, as opposed to the more generic Into. Currently, though, PoolReturnRead has this type constraint:

```
impl<R: Read + Sized + Into<Stream>> PoolReturnRead<R> {
```

Introduced in #31 / #68.

I think if we want to refactor this, it should be a separate PR.

Reply:
Sounds good!